typescript/src/schema/proto.ts

```diff
@@ -576,10 +595,120 @@ export function getHeaderRequestParams(
+// are in a sorted order (e.g. all the annotations for a single parameter are next to each other).
+
+export function getDynamicHeaderRequestParams(
+  rules: protos.google.api.IRoutingParameter[] | null
+) {
+  const params: string[][][] = [[]];
+  let counterOfParameters = 0;
+  let counterOfRules = 0;
+  const setOfParameters = new Set();
+  // if there are no rules or the rules are empty or there is a malformed routing rule, return empty array
+  try {
+    if (
+      rules === null ||
+      rules.length === 0 ||
+      getSingleRoutingHeaderParam(rules[counterOfRules]).length < 1
+    ) {
+      return [];
+    }
+  } catch (err) {
+    return [];
+  }
+  // We need to know the number of parameters we are trying to return
+  while (counterOfRules < rules.length) {
+    try {
+      setOfParameters.add(
+        getSingleRoutingHeaderParam(rules[counterOfRules])[1]
+      );
+      // Adding newer items to front of array due to "last one wins"
+      params[0].unshift(getSingleRoutingHeaderParam(rules[counterOfRules]));
+    } catch (err) {
+      return [];
+    }
+    counterOfRules++;
+  }
+  const numberOfParameters = setOfParameters.size;
+  // If there is only one parameter, then the array contains all the rules
+  if (numberOfParameters === 1) {
+    return params;
+  } else {
+    // Clear first array of params and counter
+    counterOfRules = 0;
+    params[0] = [];
+    while (counterOfRules < rules.length) {
+      // First rule to be added to first array
+      if (counterOfRules === 0) {
+        params[counterOfParameters].unshift(
+          getSingleRoutingHeaderParam(rules[counterOfRules])
+        );
+        counterOfRules++;
+      }
+      // All the rules for the same parameter go into a single array.
+      else if (
+        getSingleRoutingHeaderParam(rules[counterOfRules])[1] ===
+        getSingleRoutingHeaderParam(rules[counterOfRules - 1])[1]
+      ) {
+        // initialize next array
+        params[counterOfParameters].unshift(
+          getSingleRoutingHeaderParam(rules[counterOfRules])
+        );
+        counterOfRules++;
+      } else {
+        counterOfParameters++;
+        params[counterOfParameters] = [];
+        params[counterOfParameters].unshift(
+          getSingleRoutingHeaderParam(rules[counterOfRules])
+        );
+        counterOfRules++;
+      }
+    }
+    return params;
+  }
+}
+
+// This parses a single Routing Parameter and returns an array with the first element being the name of the
+// field the header should retrieve from the message, the second element being the name of the field the header should send,
+// the third element is the full regex of the path template that the message field should match,
+// and the last element is the named regex capture of the value of the field.
+export function getSingleRoutingHeaderParam(
+  rule: protos.google.api.IRoutingParameter
+) {
+  const mapOfRoutingParameters: string[] = [];
```

Comment:
For better readability, we could data like `map`, `Tuple`, `Interface` to store the data instead of use array[0], array[1], array[2]...
Use Tuple example:
```
interface MapRoutingParam {
 field: string,
 nameSegment: string,
....
}
```